File: backend/app/schemas/generated_artifact.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class GeneratedArtifactCreate(BaseModel):
    artifact_type: str
    title: str
    resume_id: int | None = None
    job_posting_id: int | None = None
    application_record_id: int | None = None
    content_text: str | None = None
    content_json: dict[str, Any] | None = None
    status: str = "draft"
    generator_type: str = "manual"


class GeneratedArtifactUpdate(BaseModel):
    artifact_type: str | None = None
    resume_id: int | None = None
    job_posting_id: int | None = None
    application_record_id: int | None = None
    title: str | None = None
    content_text: str | None = None
    content_json: dict[str, Any] | None = None
    status: str | None = None
    generator_type: str | None = None


class GeneratedArtifactListItem(BaseModel):
    id: int
    artifact_type: str
    resume_id: int | None
    job_posting_id: int | None
    application_record_id: int | None
    title: str
    status: str
    generator_type: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class GeneratedArtifactRead(GeneratedArtifactListItem):
    content_text: str | None
    content_json: dict[str, Any] | None

File: frontend/src/types/generated_artifact.ts
import type { ISODateString, JsonObject } from "./common";

export interface GeneratedArtifactListItem {
  id: number;
  artifact_type: string;
  resume_id: number | null;
  job_posting_id: number | null;
  application_record_id: number | null;
  title: string;
  status: string;
  generator_type: string;
  created_at: ISODateString;
  updated_at: ISODateString;
}

export interface GeneratedArtifact extends GeneratedArtifactListItem {
  content_text: string | null;
  content_json: JsonObject | null;
}

export interface GeneratedArtifactCreate {
  artifact_type: string;
  title: string;
  resume_id?: number | null;
  job_posting_id?: number | null;
  application_record_id?: number | null;
  content_text?: string | null;
  content_json?: JsonObject | null;
  status?: string;
  generator_type?: string;
}

export type GeneratedArtifactUpdate = Partial<GeneratedArtifactCreate>;

export interface CoverLetterGenerateRequest {
  resume_id: number;
  job_posting_id: number;
  application_record_id?: number | null;
  language_mode?: "zh" | "bilingual";
}

export interface InterviewPrepGenerateRequest {
  resume_id: number;
  job_posting_id: number;
  application_record_id?: number | null;
}

export type ArtifactFeedbackType =
  | "accepted"
  | "edited_then_used"
  | "rejected"
  | "saved_for_later";

export interface ArtifactFeedbackCreate {
  feedback_type: ArtifactFeedbackType;
  note?: string | null;
}

export interface ArtifactFeedback {
  id: number;
  generated_artifact_id: number;
  feedback_type: ArtifactFeedbackType;
  note: string | null;
  created_at: ISODateString;
}
